File: Strata/trunk/java/components/StrataCommon/src/strata1/common/producerconsumer/DisruptorDispatcher.java
// ##########################################################################
// # File Name:	DisruptorDispatcher.java
// #
// # Copyright:	2014, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataCommon Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataCommon Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataCommon
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata1.common.producerconsumer;

import com.lmax.disruptor.RingBuffer;
import com.lmax.disruptor.dsl.ProducerType;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

/****************************************************************************
 * 
 * @author 		
 *     Sapientia Systems
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 * @param <T>
 */
public
class DisruptorDispatcher<T>
    extends    AbstractDispatcher<T>
    implements IDispatcher<T>
{
    private final PriorityDisruptor<Event<T>> itsDisruptor;
    private final AtomicBoolean               itsRoutingIndicator;

    /************************************************************************
     * Creates a new {@code DisruptorDispatcher}. 
     *
     */
    protected 
    DisruptorDispatcher(
        IEventFactory<T> factory,
        int              numPriorities,
        int              bufferSize)
    {
        itsDisruptor = 
            new PriorityDisruptor<Event<T>>(
                factory,
                numPriorities,
                bufferSize,
                ProducerType.MULTI);
        itsDisruptor.handleExceptionsWith( 
            new DisruptorExceptionHandler(
                new PrintStackTraceExceptionHandler()) );
        
        itsRoutingIndicator = new AtomicBoolean(false);
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void
    route(int priority,T payload)
    {
        RingBuffer<Event<T>> buffer = itsDisruptor.getRingBuffer(priority);
        long                 sequence = buffer.next();
        Event<T>             event    = buffer.get( sequence );
        
        event.setKind( DispatchKind.ROUTE );
        event.setPayload( payload );        
        buffer.publish(sequence);
    }
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void
    broadcast(int priority,T payload)
    {
        RingBuffer<Event<T>> buffer = itsDisruptor.getRingBuffer(priority);
        long                 sequence = buffer.next();
        Event<T>             event    = buffer.get( sequence );
        
        event.setKind( DispatchKind.BROADCAST );
        event.setPayload( payload );        
        buffer.publish(sequence);
    }
    
    /************************************************************************
     * {@inheritDoc} 
     */
    @SuppressWarnings("unchecked")
    @Override
    public void 
    startDispatching()
    {
        List<DisruptorEventHandler<T>> eventHandlers = 
            new ArrayList<DisruptorEventHandler<T>>();
        
        for (List<IConsumer<T>> consumers:getConsumers().getValues())
        {
            long index       = 0;
            long cardinality = consumers.size();
            
            for (IConsumer<T> consumer:consumers)
                eventHandlers.add( 
                    createHandler(
                        index++,
                        cardinality,
                        consumer ) );
        }
        
        itsDisruptor.handleEventsWith( 
            eventHandlers.toArray( 
                new DisruptorEventHandler[eventHandlers.size()] ) );
        
        itsDisruptor.start();
        itsRoutingIndicator.set( true );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public void 
    stopDispatching()
    {
        itsDisruptor.shutdown();
        itsRoutingIndicator.set( false );
    }

    /************************************************************************
     * {@inheritDoc} 
     */
    @Override
    public boolean 
    isDispatching()
    {
        return itsRoutingIndicator.get();
    }
    
    /************************************************************************
     *  
     *
     * @param index
     * @param cardinality
     * @param consumer
     * @return
     */
    protected DisruptorEventHandler<T>
    createHandler(long index,long cardinality,IConsumer<T> consumer)
    {
        return new DisruptorEventHandler<T>(index,cardinality,consumer);
    }
}
// ##########################################################################

File: csharp/components/StrataCommon/ProducerConsumer/DisruptorDispatcher.cs
//  ##########################################################################
//  # File Name: DisruptorDispatcher.cs
//  # Copyright: 2016, Sapientia Systems, LLC.
//  ##########################################################################

using Disruptor;
using Disruptor.Dsl;
using Disruptor.Scheduler;
using Strata.Common.ProducerConsumer;
using Strata.Common.Utility;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Strata.Common.ProducerConsumer
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// 
    /// </summary>
    ///  
    public
    class DisruptorDispatcher<T>:
        AbstractDispatcher<T>
        where T : class
    {
        private Disruptor<Event<T>> itsDisruptor;
        private AtomicBoolean       itsDispatchingIndicator;

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public
        DisruptorDispatcher(Func<Event<T>> eventFactory, int bufferSize):
            this(eventFactory, bufferSize, 8) {}

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        public
        DisruptorDispatcher(
            Func<Event<T>> eventFactory,
            int            bufferSize,
            int            numThreads)
        {
            itsDisruptor =
                new Disruptor<Event<T>>(
                    eventFactory,
                    new MultiThreadedLowContentionClaimStrategy(bufferSize),
                    new BusySpinWaitStrategy(),
                    new RoundRobinThreadAffinedTaskScheduler(numThreads));

            itsDisruptor.HandleExceptionsWith(
                new DisruptorExceptionHandler(
                    new PrintStackTraceExceptionHandler()));

            itsDispatchingIndicator = new AtomicBoolean(false);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        Route(T payload)
        {
            RingBuffer<Event<T>> buffer   = itsDisruptor.RingBuffer;
            long                 sequence = buffer.Next();
            Event<T>             evnt     = buffer[sequence];

            evnt.Kind = DispatchKind.ROUTE;
            evnt.Payload = payload;
            buffer.Publish(sequence);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        Broadcast(T payload)
        {
            RingBuffer<Event<T>> buffer   = itsDisruptor.RingBuffer;
            long                 sequence = buffer.Next();
            Event<T>             evnt     = buffer[sequence];

            evnt.Kind = DispatchKind.BROADCAST;
            evnt.Payload = payload;
            buffer.Publish(sequence);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        StartDispatching()
        {
            IList<IEventHandler<Event<T>>> eventHandlers =
                new List<IEventHandler<Event<T>>>();

            foreach (
                IList<IConsumer<T>> consumers
                    in GetConsumers().GetValues())
            {
                long index = 0;
                long cardinality = consumers.Count;

                foreach (IConsumer<T> consumer in consumers)
                    eventHandlers.Add(
                        CreateHandler(
                            index++,
                            cardinality,
                            consumer));
            }

            itsDisruptor.HandleEventsWith(eventHandlers.ToArray());
            itsDisruptor.Start();
            itsDispatchingIndicator.Set(true);

        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override void
        StopDispatching()
        {
            itsDisruptor.Shutdown();
            itsDispatchingIndicator.Set(false);
        }

        //////////////////////////////////////////////////////////////////////
        /// <inheritDoc/>
        /// 
        public override bool
        IsDispatching()
        {
            return itsDispatchingIndicator.Get();
        }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// </summary>
        /// 
        private DisruptorEventHandler<T>
        CreateHandler(long index, long cardinality, IConsumer<T> consumer)
        {
            return
                new DisruptorEventHandler<T>(
                    index,
                    cardinality,
                    consumer);
        }

    }
}

//  ##########################################################################
